blaze-server/src/main/scala/org/http4s/blaze/server/BlazeServerBuilder.scala

```diff
@@ -269,39 +268,34 @@ class BlazeServerBuilder[F[_]] private (
     def requestAttributes(secure: Boolean, optionalSslEngine: Option[SSLEngine]): () => Vault =
       (conn.local, conn.remote) match {
         case (local: InetSocketAddress, remote: InetSocketAddress) =>
-          () =>
+          () => {
+            val connection = Request.Connection(
+              local = SocketAddress(
+                IpAddress.fromBytes(local.getAddress.getAddress).get,
+                Port.fromInt(local.getPort).get,
+              ),
+              remote = SocketAddress(
+                IpAddress.fromBytes(remote.getAddress.getAddress).get,
+                Port.fromInt(remote.getPort).get,
+              ),
+              secure = secure,
+            )
+
+            // Create SSLSession object only for https requests and if current SSL session is not empty.
+            // Here, each condition is checked inside a "flatMap" to handle possible "null" values
+            def secureSession: Option[SecureSession] =
+              for {
+                engine <- optionalSslEngine
+                session <- Option(engine.getSession)
+                hex <- Option(session.getId).map(ByteVector(_).toHex)
+                cipher <- Option(session.getCipherSuite)
+              } yield SecureSession(hex, cipher, deduceKeyLength(cipher), getCertChain(session))
```

Comment:
Does Cats do anything fancy to optimize away intermediate options in the applicative style?  I don't think so, but it certainly can't monadically.

Reply:
I had not bothered to look at these. Since these are `Option`, each line is just a comparison to `null`, so it does not seem to be the kind of thing that applicative or monadic application would optimise or parallelise.  Furthermore, note that the last 2 lines (`hex` and `cipher` depend on the result of the second line.